File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.py
class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        ROWS = len(heights)
        COLS = len(heights[0])
        DIRECTIONS = [ [1, 0], [-1, 0], [0, 1], [0, -1] ]
        pacific = set()
        atlantic = set()

        def dfs(r, c, ocean):
            ocean.add((r, c))
            for dr, dc in DIRECTIONS:
                nr = dr + r
                nc = dc + c

                if 0 <= nr < ROWS and 0 <= nc < COLS and heights[nr][nc] >= heights[r][c] and (nr, nc) not in ocean:
                    ocean.add((nr, nc))
                    dfs(nr, nc, ocean) 

        # first and last row 
        for c in range(COLS):
            dfs(0, c, pacific)
            dfs(ROWS-1, c, atlantic)

        # first and last col 
        for r in range(ROWS):
            dfs(r, 0, pacific)
            dfs(r, COLS-1, atlantic)

        ans = []
        for r in range(ROWS):
            for c in range(COLS):
                if (r, c) in pacific and (r, c) in atlantic:
                    ans.append([r, c])

        return ans

File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.js
/**
 * @param {number[][]} heights
 * @return {number[][]}
 */
var pacificAtlantic = function(heights) {
    const ROWS = heights.length;
    const COLS = heights[0].length;
    
    let pSet = new Set();
    let aSet = new Set();
    
    for( let r=0; r < ROWS; r++ ) {
        dfs( r, 0, heights[r][0], pSet );
        dfs( r, COLS-1, heights[r][COLS-1], aSet );
    }
    
    for( let c=0; c < COLS; c++ ) {
        dfs( 0, c, heights[0][c], pSet );
        dfs( ROWS-1, c, heights[ROWS-1][c], aSet );
    }
    
    let result = [];
    for( let r=0; r<ROWS; r++ ) {
        for( let c=0; c<COLS; c++ ) {
            if( pSet.has(`${r}-${c}`) && aSet.has(`${r}-${c}`) )
                result.push([r, c]);
        }
    }
    
    return result;
    
    
    function dfs( r, c, prev, set ) {
        if( r < 0 || c < 0 || r >= ROWS || c >= COLS || set.has(`${r}-${c}`) || heights[r][c] < prev  )
            return;
        
        set.add(`${r}-${c}`);
        dfs(r+1, c, heights[r][c], set);
        dfs(r-1, c, heights[r][c], set);
        dfs(r, c+1, heights[r][c], set);
        dfs(r, c-1, heights[r][c], set);
    }
};